src/main/java/io/cdap/plugin/ariba/source/AribaInputFormat.java

```diff
@@ -0,0 +1,139 @@
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.JobContext;
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Input format class to configure splits and record reader
+ */
+public class AribaInputFormat extends InputFormat<NullWritable, StructuredRecord> {
+
+  private static final Gson GSON = new Gson();
+  private static final Logger LOG = LoggerFactory.getLogger(AribaInputFormat.class);
+  private final List<InputSplit> resultSplits = new ArrayList<>();
+  private final ObjectMapper objectMapper = new ObjectMapper();
+
+  @Override
+  public List<InputSplit> getSplits(JobContext jobContext) throws IOException {
+    AribaPluginConfig pluginConfig = getPluginConfig(jobContext);
+    AribaServices aribaServices = new AribaServices(pluginConfig);
+    boolean previewEnabled = Boolean.parseBoolean(jobContext.getConfiguration().
+                                                    get(ResourceConstants.IS_PREVIEW_ENABLED));
+    createJob(pluginConfig, aribaServices, previewEnabled, null);
+    return resultSplits;
+  }
+
+  @Override
+  public RecordReader<NullWritable, StructuredRecord> createRecordReader(InputSplit inputSplit, TaskAttemptContext
+    taskAttemptContext) throws IOException {
+    AribaPluginConfig pluginConfig = getPluginConfig(taskAttemptContext);
+    AribaServices aribaServices = new AribaServices(pluginConfig);
+    Schema outputSchema = Schema.parseJson(taskAttemptContext.getConfiguration().get(ResourceConstants.OUTPUT_SCHEMA));
+    return new AribaRecordReader(aribaServices, outputSchema);
+  }
+
+  private AribaPluginConfig getPluginConfig(JobContext taskAttemptContext) {
+    return GSON.fromJson(taskAttemptContext.getConfiguration().
+                           get(ResourceConstants.ARIBA_PLUGIN_PROPERTIES), AribaPluginConfig.class);
+  }
+
+  /**
+   *  call for create Job
+   *
+   * @param pluginConfig  Plugin config
+   * @param aribaServices  Ariba services
+   * @param isPreviewEnabled Is preview enabled
+   * @param pageToken    Page token
+   * @throws IOException
+   */
+  private void createJob(AribaPluginConfig pluginConfig,
+                         AribaServices aribaServices, boolean isPreviewEnabled,
+                         String pageToken) throws IOException {
+    JsonNode jobData;
+    String jobId;
+
+    try {
+      JsonNode createJobResponse = aribaServices.createJob(pluginConfig,
+                                                           aribaServices.getAccessToken(), pageToken);
+      jobId = createJobResponse.get(ResourceConstants.JOB_ID).asText();
+    } catch (AribaException | InterruptedException | IOException exception) {
+      throw new IOException(exception.getMessage(), exception);
+    }
+
+    while (true) {
```

Comment:
Avoid infinite loops. Retry upto some MAX_RETRIES attempts, and then throw an exception if the job hasn't completed successfully.

Reply:
This gets completed in around 10 min or max 15 min which we have noticed. It's sure that this job will not run infinite.
Do you still suggest to add MAX_RETRIES?

Reply:
yes, we shouldn't have infinite loops